File: server/dotnet/Controllers/ServerController.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

namespace dotnet.Controllers
{
    public class ServerController(TemplatePathResolver _templatePathResolver, IHttpClientFactory _httpClientFactory) : Controller
    {
        private static readonly string _paypalApiBaseUrl = Environment.GetEnvironmentVariable("PAYPAL_API_BASE_URL") ?? "https://api-m.sandbox.paypal.com";
        private static readonly string _paypalSdkBaseUrl = Environment.GetEnvironmentVariable("PAYPAL_SDK_BASE_URL") ?? "https://www.sandbox.paypal.com";
        private static readonly string _paypalClientId = Environment.GetEnvironmentVariable("PAYPAL_CLIENT_ID") ?? "";
        private static readonly string _paypalClientSecret = Environment.GetEnvironmentVariable("PAYPAL_CLIENT_SECRET") ?? "";
        private static readonly string _paypalMerchantId = Environment.GetEnvironmentVariable("PAYPAL_MERCHANT_ID") ?? "";
        private static readonly string _paypalBnCode = Environment.GetEnvironmentVariable("PAYPAL_BN_CODE") ?? "";
        private static readonly string _domains = Environment.GetEnvironmentVariable("DOMAINS") ?? "";

        /* ######################################################################
         * Token generation helpers
         * ###################################################################### */
        
        private static string GetAuthAssertionToken(string clientId, string merchantId)
        {
            var header = new { alg = "none" };
            var body = new { iss = clientId, payer_id = merchantId };
            var signature = "";

            var headerEncoded = Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(header)));
            var bodyEncoded = Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(body)));
            var jwtParts = new[] { headerEncoded, bodyEncoded, signature };

            var authAssertion = string.Join('.', jwtParts);

            return authAssertion;
        }

        private static async Task<string> GetClientTokenAsync(IHttpClientFactory httpClientFactory)
        {
            try
            {
                var httpClient = httpClientFactory.CreateClient();
                var auth = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{_paypalClientId}:{_paypalClientSecret}"));
                var request = new HttpRequestMessage(HttpMethod.Post, $"{_paypalApiBaseUrl}/v1/oauth2/token");
                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", auth);
                request.Content = new FormUrlEncodedContent([
                    new KeyValuePair<string, string>("grant_type", "client_credentials"),
                    new KeyValuePair<string, string>("response_type", "client_token"),
                    new KeyValuePair<string, string>("intent", "sdk_init"),
                    new KeyValuePair<string, string>("domains[]", _domains)
                ]);

                if (!string.IsNullOrEmpty(_paypalMerchantId))
                {
                    var authAssertionToken = GetAuthAssertionToken(_paypalClientId, _paypalMerchantId);
                    request.Headers.Add("PayPal-Auth-Assertion", authAssertionToken);
                }

                var response = await httpClient.SendAsync(request);
                response.EnsureSuccessStatusCode();
                var content = await response.Content.ReadAsStringAsync();
                var jsonDocument = JsonDocument.Parse(content);
                var clientToken = jsonDocument.RootElement.GetProperty("access_token").GetString();

                return clientToken ?? "";
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine(ex);
                return "";
            }
        }

        private async Task<string> GenerateAccessToken()
        {
            var httpClient = _httpClientFactory.CreateClient();
            if (string.IsNullOrEmpty(_paypalClientId) || string.IsNullOrEmpty(_paypalClientSecret))
            {
                throw new Exception("Missing API credentials");
            }

            var url = $"{_paypalApiBaseUrl}/v1/oauth2/token";
            var auth = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{_paypalClientId}:{_paypalClientSecret}"));

            var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Basic", auth);
            request.Content = new FormUrlEncodedContent([
                new KeyValuePair<string, string>("grant_type", "client_credentials")
            ]);

            if (!string.IsNullOrEmpty(_paypalMerchantId))
            {
                request.Headers.Add("PayPal-Partner-Attribution-ID", _paypalBnCode);
                request.Headers.Add("PayPal-Auth-Assertion", GetAuthAssertionToken(_paypalClientId, _paypalMerchantId));
            }

            var response = await httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();

            var content = await response.Content.ReadAsStringAsync();
            using var document = JsonDocument.Parse(content);
            var root = document.RootElement;
            return root.GetProperty("access_token").GetString() ?? "";
        }

        /* ######################################################################
         * Serve checkout page
         * ###################################################################### */
        
        private static string GetPayPalSdkUrl()
        {
            if (string.IsNullOrEmpty(_paypalClientId))
            {
                throw new InvalidOperationException("Missing PAYPAL_CLIENT_ID");
            }

            var sdkUrl = new UriBuilder(new Uri(new Uri(_paypalSdkBaseUrl), "/sdk/js"));
            var sdkParams = new StringBuilder();
            sdkParams.Append($"client-id={_paypalClientId}&components=buttons,fastlane");

            sdkUrl.Query = sdkParams.ToString();

            return sdkUrl.ToString();
        }

        [HttpGet("/")]
        public async Task<IActionResult> RenderCheckout()
        {
            bool isFlexibleIntegration = HttpContext.Request.Query.Any(param => param.Key == "flexible");

            string sdkUrl = GetPayPalSdkUrl();
            string clientToken = await GetClientTokenAsync(_httpClientFactory);

            var locals = new Dictionary<string, string>
        {
            { "title", "Fastlane - PayPal Integration" + (isFlexibleIntegration ? " (Flexible)" : "") },
            { "prerequisiteScripts", $@"
                <script
                    src=""{sdkUrl}""
                    data-sdk-client-token=""{clientToken}""
                    defer
                ></script>
            " },
            { "initScriptPath", isFlexibleIntegration ? "init-fastlane-flexible.js" : "init-fastlane.js" },
            { "stylesheetPath", "styles.css" }
        };

            string renderedHtml = await _templatePathResolver.RenderTemplateAsync(isFlexibleIntegration, locals);

            return Content(renderedHtml, "text/html", Encoding.UTF8);
        }

       /* ######################################################################
        * Process transactions
        * ###################################################################### */

        [HttpPost("transaction")]
        public async Task<IActionResult> CreateOrder([FromBody] CreateOrderRequest request)
        {
            try
            {
                var httpClient = _httpClientFactory.CreateClient("NoSSL");
                var accessToken = await GenerateAccessToken();
                var url = $"{_paypalApiBaseUrl}/v2/checkout/orders";

                var fullName = request?.ShippingAddress?.Name?.FullName;
                var countryCode = request?.ShippingAddress?.PhoneNumber?.CountryCode;
                var nationalNumber = request?.ShippingAddress?.PhoneNumber?.NationalNumber;

                var payload = new
                {
                    intent = "CAPTURE",
                    payment_source = new
                    {
                        card = new
                        {
                            single_use_token = request?.PaymentToken?.Id
                        }
                    },
                    purchase_units = new[]
                    {
                    new
                    {
                        amount = new
                        {
                            currency_code = "USD",
                            value = "110.00"
                        },
                        shipping = request?.ShippingAddress == null ? null : new
                        {
                            type = "SHIPPING",
                            name = !string.IsNullOrEmpty(fullName) ? new { full_name = fullName } : null,
                            company_name = !string.IsNullOrEmpty(request.ShippingAddress?.CompanyName) ? request.ShippingAddress?.CompanyName : null,
                            address = request.ShippingAddress != null ? new
                            {
                                address_line_1 = request.ShippingAddress.Address?.AddressLine1,
                                address_line_2 = request.ShippingAddress.Address?.AddressLine2,
                                admin_area_2 = request.ShippingAddress.Address?.AdminArea2,
                                admin_area_1 = request.ShippingAddress.Address?.AdminArea1,
                                postal_code = request.ShippingAddress.Address?.PostalCode,
                                country_code = request.ShippingAddress.Address?.CountryCode
                            } : null,
                            phone_number = !string.IsNullOrEmpty(countryCode) && !string.IsNullOrEmpty(nationalNumber) ?  new {
                                country_code = countryCode,
                                national_number = nationalNumber
                            } : null
                        }
                    }
                }
                };

                var requestMessage = new HttpRequestMessage(HttpMethod.Post, url);
                requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
                requestMessage.Headers.Add("PayPal-Request-Id", DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString());

                var content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");
                requestMessage.Content = content;

                var response = await httpClient.SendAsync(requestMessage);
                var responseContent = await response.Content.ReadAsStringAsync();

                return StatusCode((int)response.StatusCode, new { result = JsonDocument.Parse(responseContent).RootElement });
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
                return StatusCode(500, new { error = ex.Message });
            }
        }

        [HttpGet("/sdk/url")]
        public IActionResult GetSdkUrl()
        {
            var sdkUrl = GetPayPalSdkUrl();
            return new OkObjectResult(new { url = sdkUrl });
        }

        [HttpGet("/sdk/client-token")]
        public async Task<IActionResult> GetClientToken()
        {
            var clientToken = await GetClientTokenAsync(_httpClientFactory);
            return new OkObjectResult(new { clientToken });
        }
    }

   /* ######################################################################
    * Models
    * ###################################################################### */

    public class CreateOrderRequest
    {
        public PaymentToken? PaymentToken { get; set; }
        public ShippingAddress? ShippingAddress { get; set; }
    }

    public class PaymentToken
    {
        public string? Id { get; set; }
    }

    public class ShippingAddress
    {
        public Name? Name { get; set; }
        public string? CompanyName { get; set; }
        public Address? Address { get; set; }
        public PhoneNumber? PhoneNumber { get; set; }
    }

    public class Name
    {
        public string? FullName { get; set; }
    }

    public class Address
    {
        public string? AddressLine1 { get; set; }
        public string? AddressLine2 { get; set; }
        public string? AdminArea2 { get; set; }
        public string? AdminArea1 { get; set; }
        public string? PostalCode { get; set; }
        public string? CountryCode { get; set; }
    }

    public class PhoneNumber
    {
        public string? CountryCode { get; set; }
        public string? NationalNumber { get; set; }
    }

    public class AccessTokenResponse
    {
        public string? access_token { get; set; }
    }
}

File: server/php/src/Controller/ServerController.php
<?php

namespace App\Controller;

use Mustache_Engine;
use Error;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpClient\HttpClient;
use Symfony\Component\HttpFoundation\JsonResponse;

/* ######################################################################
 * Token generation helpers
 * ###################################################################### */

class SdkScriptHelpers
{
    public static function getPayPalSDKUrl(): string
    {
        $sdkUrl = Env::get("PAYPAL_SDK_BASE_URL") . "/sdk/js";

        $sdkQueryParams = http_build_query([
            "client-id" => Env::get("PAYPAL_CLIENT_ID"),
            "components" => "buttons,fastlane",
        ]);

        $fullUrl = $sdkUrl . "?" . $sdkQueryParams;

        return $fullUrl;
    }

    public static function getClientToken(): string
    {
        if (
            !Env::get("PAYPAL_CLIENT_ID") ||
            !Env::get("PAYPAL_CLIENT_SECRET")
        ) {
            throw new Error("Missing API credentials");
        }

        $httpClient = HttpClient::create();

        $url = Env::get("PAYPAL_API_BASE_URL") . "/v1/oauth2/token";

        $authKey = base64_encode(
            Env::get("PAYPAL_CLIENT_ID") .
                ":" .
                Env::get("PAYPAL_CLIENT_SECRET")
        );

        $headers = [
            "Authorization" => "Basic " . $authKey,
            "Content-Type" => "application/x-www-form-urlencoded",
        ];

        if (Env::get("PAYPAL_MERCHANT_ID")) {
            $headers["PayPal-Auth-Assertion"] = Auth::getAuthAssertionToken(
                Env::get("PAYPAL_CLIENT_ID"),
                Env::get("PAYPAL_MERCHANT_ID")
            );
        }

        $response = $httpClient
            ->request("POST", $url, [
                "headers" => $headers,
                "body" => [
                    "grant_type" => "client_credentials",
                    "response_type" => "client_token",
                    "intent" => "sdk_init",
                    "domains[]" => Env::get("DOMAINS"),
                ],
            ])
            ->toArray();

        return $response["access_token"];
    }

    public static function getAccessToken(): string
    {
        if (
            !Env::get("PAYPAL_CLIENT_ID") ||
            !Env::get("PAYPAL_CLIENT_SECRET")
        ) {
            throw new Error("Missing API credentials");
        }

        $httpClient = HttpClient::create();

        $url = Env::get("PAYPAL_API_BASE_URL") . "/v1/oauth2/token";

        $authKey = base64_encode(
            Env::get("PAYPAL_CLIENT_ID") .
                ":" .
                Env::get("PAYPAL_CLIENT_SECRET")
        );

        $headers = [
            "Authorization" => "Basic " . $authKey,
            "Content-Type" => "application/x-www-form-urlencoded",
        ];

        if (Env::get("PAYPAL_MERCHANT_ID")) {
            $headers["PayPal-Partner-Attribution-ID"] = Env::get(
                "PAYPAL_BN_CODE"
            );
            $headers["PayPal-Auth-Assertion"] = Auth::getAuthAssertionToken(
                Env::get("PAYPAL_CLIENT_ID"),
                Env::get("PAYPAL_MERCHANT_ID")
            );
        }

        $response = $httpClient
            ->request("POST", $url, [
                "headers" => $headers,
                "body" => [
                    "grant_type" => "client_credentials",
                ],
            ])
            ->toArray();

        return $response["access_token"];
    }
}

class ServerController extends AbstractController
{
    private $mustache;

    public function __construct()
    {
        $this->mustache = new Mustache_Engine(["entity_flags" => ENT_QUOTES]);
    }

    /* ######################################################################
     * Serve checkout page
     * ###################################################################### */

    public function index(Request $request)
    {
        $isFlexibleIntegration = $request->query->get("flexible", false);

        $sdkUrl = SdkScriptHelpers::getPayPalSDKUrl();
        $clientToken = SdkScriptHelpers::getClientToken();

        $locals = [
            "title" =>
                "Fastlane - PayPal Integration" .
                ($isFlexibleIntegration ? " (Flexible)" : ""),
            "prerequisiteScripts" => "
                <script src='$sdkUrl'
                    data-sdk-client-token='$clientToken'
                    defer
                ></script>
            ",
            "initScriptPath" => $isFlexibleIntegration
                ? "init-fastlane-flexible.js"
                : "init-fastlane.js",
            "stylesheetPath" => "styles.css",
        ];

        $htmlTemplate = file_get_contents(
            __DIR__ .
                "/../../../shared/views/" .
                ($isFlexibleIntegration
                    ? "checkout-flexible.html"
                    : "checkout.html")
        );

        $template = $this->mustache->render($htmlTemplate, $locals);

        return new Response($template);
    }

    /* ######################################################################
     * Process transactions
     * ###################################################################### */

    public function createTransaction(Request $request)
    {
        $data = $request->toArray();

        $url = Env::get("PAYPAL_API_BASE_URL") . "/v2/checkout/orders";

        $accessToken = SdkScriptHelpers::getAccessToken();

        $headers = [
            "PayPal-Request-Id" => (string) (time() * 1000),
            "Authorization" => "Bearer " . $accessToken,
            "Content-Type" => "application/json",
        ];

        $httpClient = HttpClient::create();

        $body = [
            "intent" => "CAPTURE",
            "payment_source" => [
                "card" => [
                    "single_use_token" => $data["paymentToken"]["id"],
                ],
            ],
            "purchase_units" => [
                [
                    "amount" => [
                        "currency_code" => "USD",
                        "value" => "110.00",
                    ],
                ],
            ],
        ];

        if (array_key_exists("shippingAddress", $data)) {
            $shippingAddress = $data["shippingAddress"];

            $fullName = isset($shippingAddress["name"]["fullName"])
                ? $shippingAddress["name"]["fullName"]
                : null;

            $companyName = isset($shippingAddress["companyName"])
                ? $shippingAddress["companyName"]
                : null;
            $countryCode = isset($shippingAddress["phoneNumber"]["countryCode"])
                ? $shippingAddress["phoneNumber"]["countryCode"]
                : null;
            $nationalNumber = isset(
                $shippingAddress["phoneNumber"]["nationalNumber"]
            )
                ? $shippingAddress["phoneNumber"]["nationalNumber"]
                : null;

            $body["purchase_units"][0]["shipping"] = [
                "type" => "SHIPPING",
                "name" => $fullName ? ["full_name" => $fullName] : null,
                "company_name" => !empty($companyName) ? $companyName : null,
                "address" => [
                    "address_line_1" => isset(
                        $shippingAddress["address"]["addressLine1"]
                    )
                        ? $shippingAddress["address"]["addressLine1"]
                        : null,
                    "address_line_2" => isset(
                        $shippingAddress["address"]["addressLine2"]
                    )
                        ? $shippingAddress["address"]["addressLine2"]
                        : null,
                    "admin_area_2" => isset(
                        $shippingAddress["address"]["adminArea2"]
                    )
                        ? $shippingAddress["address"]["adminArea2"]
                        : null,
                    "admin_area_1" => isset(
                        $shippingAddress["address"]["adminArea1"]
                    )
                        ? $shippingAddress["address"]["adminArea1"]
                        : null,
                    "postal_code" => isset(
                        $shippingAddress["address"]["postalCode"]
                    )
                        ? $shippingAddress["address"]["postalCode"]
                        : null,
                    "country_code" => isset(
                        $shippingAddress["address"]["countryCode"]
                    )
                        ? $shippingAddress["address"]["countryCode"]
                        : null,
                ],
            ];
        }

        if ($countryCode && $nationalNumber) {
            $body["purchase_units"][0]["shipping"]["phone_number"] = [
                "country_code" => $countryCode,
                "national_number" => $nationalNumber,
            ];
        }

        $result = $httpClient
            ->request("POST", $url, [
                "headers" => $headers,
                "body" => json_encode($body),
            ])
            ->toArray();

        return new JsonResponse(["result" => $result]);
    }

    public function getSDKUrl()
    {
        $sdkUrl = SdkScriptHelpers::getPayPalSDKUrl();
        return $this->json(["url" => $sdkUrl])->setEncodingOptions(
            JSON_UNESCAPED_SLASHES
        );
    }

    public function getClientToken()
    {
        $clientToken = SdkScriptHelpers::getClientToken();
        return $this->json(["clientToken" => $clientToken]);
    }
}

class Auth
{
    public static function getAuthAssertionToken(
        string $clientId,
        string $merchantId
    ): string {
        $header = [
            "alg" => "none",
        ];

        $body = [
            "iss" => $clientId,
            "payer_id" => $merchantId,
        ];

        $signature = "";

        $jwtParts = [$header, $body, $signature];

        $authAssertion = array_map(function ($part) {
            return $part ? base64_encode(json_encode($part)) : "";
        }, $jwtParts);

        return join(".", $authAssertion);
    }
}

class Env
{
    static function get(string $key): string|null
    {
        return $_ENV[$key] ?? null;
    }
}
